signer/src/stacks/contracts.rs

```diff
@@ -987,11 +1067,110 @@ impl AsContractCall for RejectWithdrawalV1 {
+            return Err(WithdrawalRejectErrorMsg::DeployerMismatch.into_error(req_ctx, self));
+        }
+
+        // 1. The request exists. Check whether the associated withdrawal request transaction is confirmed on the canonical stacks blockchain. Fail the withdrawal request if it is not on the canonical stacks blockchain.
+        //
+        // 2. The double spend check. Check whether the qualified request ID is in the bitcoin_withdrawals_outputs table, and that any of the associated txids are confirmed on the canonical bitcoin blockchain. Fail the withdrawal request if such a transaction was found.
+
+        eprintln!("get withdrawal request report");
+        let Some(report) = ctx
+            .get_storage()
+            .get_withdrawal_request_report(
+                &req_ctx.chain_tip.block_hash,
+                &self.id.block_hash,
+                &self.id,
+                &req_ctx.origin,
+            )
+            .await
+            .map_err(|_| WithdrawalRejectErrorMsg::RequestMissing.into_error(req_ctx, self))?
+        else {
+            eprintln!("no withdrawal request report");
+            return Err(WithdrawalRejectErrorMsg::RequestMissing.into_error(req_ctx, self));
+        };
+
+        eprintln!("check withdrawal request report status");
+        match report.status {
+            WithdrawalRequestStatus::Confirmed => (),
+            WithdrawalRequestStatus::Fulfilled(_txid) => {
+                // fails #2
+                eprintln!("withdrawal request was fulfilled");
+                return Err(
+                    WithdrawalRejectErrorMsg::WithdrawalRequestFulfilled.into_error(req_ctx, self)
+                );
+            }
+            WithdrawalRequestStatus::Unconfirmed => {
+                // fails #1
+                eprintln!("withdrawal request was unconfirmed");
+                return Err(WithdrawalRejectErrorMsg::WithdrawalRequestUnconfirmed
+                    .into_error(req_ctx, self));
+            }
+        }
+
+        eprintln!("validate final request");
+        // 3. The request is final. Check that we've seen six new bitcoin blocks since observing the bitcoin anchor block associated with the Stacks block confirming the withdrawal request transaction. Fail the withdrawal request if we’ve observed less than 6 bitcoin blocks since the anchor block.
+
+        let request_block_height = report.block_height;
+
+        let Some(tip_block_hash) = ctx.get_storage().get_bitcoin_canonical_chain_tip().await?
+        else {
+            eprintln!("no bitcoin canonical chain tip");
+            return Err(WithdrawalRejectErrorMsg::NoBitcoinChainTip.into_error(req_ctx, self));
+        };
+
+        let Some(tip_bitcoin_block) = ctx.get_storage().get_bitcoin_block(&tip_block_hash).await?
+        else {
+            eprintln!("no bitcoin block for canonical chain tip");
+            return Err(WithdrawalRejectErrorMsg::MissingBitcoinBlock.into_error(req_ctx, self));
+        };
```

Comment:
Oh, no need for this, you should have what you need in the `ReqContext`.

Reply:
The chain tip in `ReqContext` is the sweep block.  In order to finalize the withdrawal request, I have to mine 6 more blocks.

Reply:
In our tests we often set it to that but generally, what is in there should be the chain tip.

Reply:
Would you prefer that I pass the `db` into `make_withdrawal_reject2` and set the real chain tip as above?

Reply:
Yeah, that's fine. Or pass it the chain tip.

Reply:
Done

Reply:
(won't resolve this until you look at it)